src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,625 @@
+  }
+  return std::string(buf);
+}
+
+void NativeModule::find_LongFuncL(
+  JNIEnv* env,
+  NativeStatus* st,
+  jobject funcObj,
+  jstring nameJ,
+  int numArgs
+) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  void* funcAddr = nullptr;
+  if (!try_load_locked()) {
+    NATIVE_ERROR(st, 1001, "ErrModuleNotFound");
+  } else {
+    auto qualName = to_qualified_name(env, key_, nameJ, numArgs, is_global_, true);
+    funcAddr = ::dlsym(is_global_ ? RTLD_DEFAULT : dlopen_handle_, qualName.c_str());
+    if (!funcAddr) {
+      fprintf(stderr, "ErrLongFuncNotFound \"%s\"\n", qualName.c_str());
+      NATIVE_ERROR(st, 1003, "ErrLongFuncNotFound dlsym(\"%s\")", qualName.c_str());
+    }
+  }
+  NativeObjPtr ptr = std::make_shared< NativeFuncObj<long> >(shared_from_this(), funcAddr);
+  init_NativePtr(env, funcObj, &ptr);
+}
+
+void NativeModule::find_PtrFuncL(
+  JNIEnv* env,
+  NativeStatus* st,
+  jobject funcObj,
+  jstring nameJ,
+  int numArgs
+) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  void* funcAddr = nullptr;
+  if (!try_load_locked()) {
+    NATIVE_ERROR(st, 1001, "ErrModuleNotFound");
+  } else {
+    auto qualName = to_qualified_name(env, key_, nameJ, numArgs, is_global_, false);
+    funcAddr = ::dlsym(is_global_ ? RTLD_DEFAULT : dlopen_handle_, qualName.c_str());
+    if (!funcAddr) {
+      fprintf(stderr, "ErrPtrFuncNotFound \"%s\"\n", qualName.c_str());
+      NATIVE_ERROR(st, 1003, "ErrPtrFuncNotFound dlsym(\"%s\")", qualName.c_str());
+    }
+  }
+  NativeObjPtr ptr = std::make_shared< NativeFuncObj<NativeObjPtr> >(shared_from_this(), funcAddr);
+  init_NativePtr(env, funcObj, &ptr);
+}
+
+// Functions implementing NativeModule native methods
+
+static NativeModule* to_NativeModule(JNIEnv* env, jobject obj) {
+  return static_cast<NativeModule*>(get_from_NativePtr(env, obj));
+}
+
+NATIVEMETHOD(void, NativeModule, nativeCtorMaster)(
+  JNIEnv* env,
+  jobject thisJ,
+  jstring optionsJ,
+  jstring sourceJ,
+  jstring includeJ,
+  jboolean force_buildJ
+) {
+  JString options(env, optionsJ);
+  JString source(env, sourceJ);
+  JString include(env, includeJ);
+  bool force_build = (force_buildJ != JNI_FALSE);
+  NativeObjPtr ptr = std::make_shared<NativeModule>(options, source, include, force_build);
+  init_NativePtr(env, thisJ, &ptr);
+}
+
+NATIVEMETHOD(void, NativeModule, nativeCtorWorker)(
+  JNIEnv* env,
+  jobject thisJ,
+  jboolean is_globalJ,
+  jstring keyJ,
+  jbyteArray binaryJ
+) {
+  bool is_global = (is_globalJ != JNI_FALSE);
+  JString key(env, keyJ);
+  NativeModulePtr mod;
+  auto iter = module_table.find(std::string(key));
+  if (iter != module_table.end()) {
+    mod = iter->second.lock(); // table holds weak_ptr, get a shared_ptr
+  }
+  if (!mod) {
+    ssize_t binary_size = env->GetArrayLength(binaryJ);
+    auto binary = env->GetByteArrayElements(binaryJ, 0);
+    mod = std::make_shared<NativeModule>(is_global, key, binary_size, binary);
+    module_table[std::string(key)] = mod;
```

Comment:
Doesn't this need to be protected?

Reply:
After looking over the code a bit more, it seems like `nativeCtorWorker` is the one and only place you need to lock.

Reply:
You're right, in trying to move lock_guard's consistently into the NativeModule methods
(with a view to making it correct when NativeModule is used from C++ without going
through the wrappers) I missed this place where we also need the big_mutex.